File: src/cli/authCommands.ts
/**
 * Auth CLI commands (Milestone 11): `zeesh login|register|logout|whoami`.
 *
 * Login/register prompt for credentials (passwords are hidden), call the
 * backend, and persist the session token in ~/.zeesh/auth.json (0600).
 * Logout invalidates the session server-side and wipes the local copy.
 * Whoami shows the authenticated identity, validating against the server
 * when reachable and degrading to the cached session when offline.
 */
import { zeeshApiUrl } from '../config/config.ts';
import { ApiClient, ApiError } from '../auth/client.ts';
import { clearSession, loadSession, saveSession, sessionExpired, type StoredSession } from '../auth/session.ts';
import { c } from './colors.ts';
import { promptHidden, promptText } from './input.ts';

/** `zeesh login [email]` */
export async function cmdLogin(arg: string): Promise<number> {
  const existing = loadSession();
  const apiUrl = existing?.apiUrl ?? zeeshApiUrl();
  console.log(c.dim(`ZEESH AI backend: ${apiUrl}`));

  const email = (arg.trim() || (await promptText(c.bold('Email: ')))).trim();
  const password = await promptHidden(c.bold('Password: '));
  if (!email || !password) {
    console.log(c.yellow('Login cancelled.'));
    return 1;
  }

  try {
    const result = await new ApiClient(apiUrl).login(email, password);
    persist(result, apiUrl);
    console.log(c.green(`Logged in as ${result.user.email}.`));
    return 0;
  } catch (err) {
    if (err instanceof ApiError && err.status === 401) {
      console.log(c.yellow('Invalid email or password. No account yet? Try "zeesh register".'));
    } else if (err instanceof ApiError && err.status === 429) {
      console.log(c.yellow(`Too many login attempts — try again in ${err.retryAfterSeconds ?? 60}s.`));
    } else if (err instanceof ApiError && err.status === 403) {
      console.log(c.yellow(err.message));
    } else {
      console.log(c.red(err instanceof Error ? err.message : 'Login failed.'));
    }
    return 1;
  }
}

/** `zeesh register [email]` — create an account (password ≥ 8 chars). */
export async function cmdRegister(arg: string): Promise<number> {
  const existing = loadSession();
  const apiUrl = existing?.apiUrl ?? zeeshApiUrl();
  console.log(c.dim(`ZEESH AI backend: ${apiUrl}`));

  const email = (arg.trim() || (await promptText(c.bold('Email: ')))).trim();
  const password = await promptHidden(c.bold('Password: '));
  if (password.length < 8) {
    console.log(c.red('Password must be at least 8 characters.'));
    return 1;
  }
  const confirm = await promptHidden(c.bold('Confirm password: '));
  if (password !== confirm) {
    console.log(c.red('Passwords do not match.'));
    return 1;
  }

  try {
    const result = await new ApiClient(apiUrl).register(email, password);
    persist(result, apiUrl);
    console.log(c.green(`Account created — logged in as ${result.user.email}.`));
    return 0;
  } catch (err) {
    if (err instanceof ApiError && err.status === 409) {
      console.log(c.yellow('An account with this email already exists. Try "zeesh login".'));
    } else if (err instanceof ApiError && err.status === 429) {
      console.log(c.yellow(`Too many attempts — try again in ${err.retryAfterSeconds ?? 60}s.`));
    } else if (err instanceof ApiError && err.status === 403) {
      console.log(c.yellow(err.message));
    } else {
      console.log(c.red(err instanceof Error ? err.message : 'Registration failed.'));
    }
    return 1;
  }
}

/** `zeesh logout` */
export async function cmdLogout(): Promise<number> {
  const session = loadSession();
  if (!session) {
    console.log(c.dim('Not logged in.'));
    return 0;
  }
  let serverOk = true;
  try {
    await new ApiClient(session.apiUrl, 3000).logout(session.token);
  } catch {
    serverOk = false; // Backend unreachable — local logout still succeeds.
  }
  clearSession();
  console.log(c.green(serverOk ? 'Logged out — local session removed.' : 'Logged out locally (backend unreachable — session may still be valid there).'));
  return 0;
}

/** `zeesh whoami` */
export async function cmdWhoami(): Promise<number> {
  const session = loadSession();
  if (!session) {
    console.log(c.dim('Not logged in. Run "zeesh login" to connect to the ZEESH AI backend.'));
    return 1;
  }

  console.log(c.bold('ZEESH AI session'));
  console.log(`  Email:     ${session.user.email}`);
  console.log(`  User ID:   ${session.user.id}`);
  console.log(`  Backend:   ${session.apiUrl}`);
  console.log(`  Expires:   ${session.expiresAt ? new Date(session.expiresAt).toLocaleString() : '—'}`);

  if (sessionExpired(session)) {
    console.log(c.yellow('  Status:    expired — run "zeesh login" again.'));
    return 1;
  }

  try {
    const user = await new ApiClient(session.apiUrl, 5000).me(session.token);
    const display = user.display_name ? ` · ${user.display_name}` : '';
    console.log(c.green(`  Status:    valid${display}`));
    return 0;
  } catch (err) {
    if (err instanceof ApiError && err.status === 401) {
      clearSession();
      console.log(c.yellow('  Status:    invalid session — run "zeesh login" again.'));
      return 1;
    }
    console.log(c.dim('  Status:    cannot reach backend (offline) — using cached session'));
    return 0;
  }
}

function persist(result: { token: string; expires_at: string; user: { id: string; email: string; display_name: string | null } }, apiUrl: string): void {
  const session: StoredSession = {
    apiUrl,
    token: result.token,
    user: { id: result.user.id, email: result.user.email, displayName: result.user.display_name },
    expiresAt: result.expires_at,
    createdAt: new Date().toISOString(),
  };
  saveSession(session);
}

File: grace/cli/auth_commands.py
"""Auth CLI commands (port of src/cli/authCommands.ts): `grace login|register|logout|whoami`.

Login/register prompt for credentials (passwords are hidden), call the
backend, and persist the session token in ~/.zeesh/auth.json (0600).
Logout invalidates the session server-side and wipes the local copy.
Whoami shows the authenticated identity, validating against the server when
reachable and degrading to the cached session when offline.
"""

from datetime import datetime, timezone

from grace.auth.client import ApiClient, ApiError
from grace.auth.session import (
    clear_session,
    load_session,
    save_session,
    session_expired,
)
from grace.cli.input import prompt_hidden, prompt_text
from grace.colors import c
from grace.config import is_local_backend, zeesh_api_url


def _backend_note(api_url: str) -> str:
    if is_local_backend(api_url):
        return f"GRACE backend: {api_url} (local dev — set ZEESH_API_URL to the deployed backend for production)"
    return f"GRACE backend: {api_url}"


def _persist(result: dict, api_url: str) -> None:
    session = {
        "apiUrl": api_url,
        "token": result["token"],
        "user": {"id": result["user"]["id"], "email": result["user"]["email"], "displayName": result["user"].get("display_name")},
        "expiresAt": result["expires_at"],
        "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    save_session(session)


def _handle_error(err: Exception, kind: str = "login") -> int:
    if isinstance(err, ApiError):
        if err.status == 401 and kind == "login":
            print(c.yellow('Invalid email or password. No account yet? Try "grace register".'))
        elif err.status == 409 and kind == "register":
            print(c.yellow('An account with this email already exists. Try "grace login".'))
        elif err.status == 429:
            print(c.yellow(f"Too many attempts — try again in {err.retry_after_seconds or 60}s."))
        elif err.status == 403:
            print(c.yellow(err.args[0] if err.args else str(err)))
        else:
            print(c.red(str(err)))
    else:
        print(c.red(str(err)))
    return 1


def cmd_login(arg: str) -> int:
    # The backend comes from configuration ONLY (ZEESH_API_URL override, else
    # the deployed production backend). A stale stored session must never pull
    # login back to an old dev backend.
    api_url = zeesh_api_url()
    print(c.dim(_backend_note(api_url)))

    email = (arg.strip() or prompt_text(c.bold("Email: "))).strip()
    password = prompt_hidden(c.bold("Password: "))
    if not email or not password:
        print(c.yellow("Login cancelled."))
        return 1

    try:
        result = ApiClient(api_url).login(email, password)
        _persist(result, api_url)
        print(c.green(f"Logged in as {result['user']['email']}."))
        return 0
    except Exception as err:
        return _handle_error(err, "login")


def cmd_register(arg: str) -> int:
    api_url = zeesh_api_url()
    print(c.dim(_backend_note(api_url)))

    email = (arg.strip() or prompt_text(c.bold("Email: "))).strip()
    password = prompt_hidden(c.bold("Password: "))
    if len(password) < 8:
        print(c.red("Password must be at least 8 characters."))
        return 1
    confirm = prompt_hidden(c.bold("Confirm password: "))
    if password != confirm:
        print(c.red("Passwords do not match."))
        return 1

    try:
        result = ApiClient(api_url).register(email, password)
        _persist(result, api_url)
        print(c.green(f"Account created — logged in as {result['user']['email']}."))
        return 0
    except Exception as err:
        return _handle_error(err, "register")


def cmd_logout() -> int:
    session = load_session()
    if not session:
        print(c.dim("Not logged in."))
        return 0
    server_ok = True
    try:
        ApiClient(session["apiUrl"], 3000).logout(session["token"])
    except Exception:
        server_ok = False  # Backend unreachable — local logout still succeeds.
    clear_session()
    print(c.green("Logged out — local session removed." if server_ok else "Logged out locally (backend unreachable — session may still be valid there)."))
    return 0


def cmd_whoami() -> int:
    session = load_session()
    if not session:
        print(c.dim('Not logged in. Run "grace login" to connect to the GRACE backend.'))
        return 1

    print(c.bold("GRACE session"))
    print(f"  Email:     {session['user']['email']}")
    print(f"  User ID:   {session['user']['id']}")
    print(f"  Backend:   {session['apiUrl']}")
    try:
        from datetime import datetime as dt

        expires = dt.fromisoformat(session.get("expiresAt", "").replace("Z", "+00:00")).strftime("%c") if session.get("expiresAt") else "—"
    except Exception:
        expires = session.get("expiresAt") or "—"
    print(f"  Expires:   {expires}")

    if session_expired(session):
        print(c.yellow('  Status:    expired — run "grace login" again.'))
        return 1

    try:
        user = ApiClient(session["apiUrl"], 5000).me(session["token"])
        display = f" · {user['display_name']}" if user.get("display_name") else ""
        print(c.green(f"  Status:    valid{display}"))
        return 0
    except ApiError as err:
        if err.status == 401:
            clear_session()
            print(c.yellow('  Status:    invalid session — run "grace login" again.'))
            return 1
        print(c.dim("  Status:    cannot reach backend (offline) — using cached session"))
        return 0
    except Exception:
        print(c.dim("  Status:    cannot reach backend (offline) — using cached session"))
        return 0
